Tests/ApplicativeRouterTests/ApplicativeRouterTests.swift

```diff
@@ -163,3 +201,25 @@ enum PostTestRoute: Equatable {
     }
   }
 }
+
+/// A struct that holds a ref tag that can be plucked from a query string, along with whatever else
+/// could get plucked from the request.
+struct RefTag<A> {
+  let ref: String?
+  let rest: A
+}
+
+/// Converts a parser of `A` into a parser of `RefTag<A>` by first trying to parse out the `A`, and then
+/// parsing a query param `ref` and storing both values in `RefTag<A>`.
+func refTag<I, A>(_ parser: Parser<I, A>) -> Parser<I, RefTag<A>> {
```

Comment:
here I use `flatMap` to take any parsers of `A` and turn it into a parser of `RefTag<A>`. it's pretty fun!

also cool to see that `I` is allowed to be free since query params can exist on requests of all methods. however, if we tried this trick on trying to always attempt plucking out some post body data we could restrict it to only `POST`s (did anyone say _csrf tokens_?!). I'm really starting to dig the phantom param for http method.